Solutions manifest

alexandre-allard opened the pull request:
**Component**: salt, solutions

**Context**: See #2422

**Summary**:
Replace `config.yaml` + `product.txt` with a single `manifest.yaml`.
Also the format has changed a bit to look more like a valid CR.

**Acceptance criteria**:
Solution workflow is still working (import, activate, ...) and the tests are green.

---

Closes: #2422

alexandre-allard committed:
salt: fix use the previously defined variable

use the variable define few lines above instead
of recomputing it.

alexandre-allard committed:
salt: add method to retrieve Solution info from manifest

this function search for manifest.yaml on the ISO
and extract the information contained in it and
return them.

Refs: #2422

alexandre-allard committed:
salt: update Solutions module to handle manifest

We know use a `manifest.yaml` instead of the
`product.txt` + the `config.yaml` to retrieve a
Solution information such as the name, the version, ...
Also rename the kind from `SolutionConfig` to `Solution`
to avoid confusion with `SolutionsConfiguration` and
because it's not a configuration file.
Note that the `manifest.yaml` file is now mandatory
unlike the old `config.yaml`.

Refs: #2422

alexandre-allard committed:
salt: remove useless variable declaration

These 2 variables were not used anywhere,
probably a remnant of the previous Solutions
version.

alexandre-allard committed:
salt: adapt Solution import for new manifest.yaml

Refs: #2422

alexandre-allard committed:
salt: adapt env preparation for new manifest.yaml

Refs: #2422

alexandre-allard committed:
example: add script to generate Solution manifest

This script will be used to generate a `manifest.yaml`
to replace the `product.txt` + `config.yaml` files:

./manifest.py  \
  --name "example-solution" \
  --annotation "solutions.metalk8s.scality.com/build-timestamp" "2020-05-19T14:51:43Z" \
  --annotation "solutions.metalk8s.scality.com/build-host" "Latitude-E7450" \
  --annotation "solutions.metalk8s.scality.com/development-release" "0" \
  --annotation "solutions.metalk8s.scality.com/display-name" "Example Solution" \
  --annotation "solutions.metalk8s.scality.com/git-revision" "2.4.3-104-g893792ca" \
  --extra-image "base-server" "0.1.0-dev" \
  --operator-image "example-solution-operator" "0.1.0-dev" \
  --ui-image "example-solution-ui" "0.1.0-dev" \
  --version "0.1.0-dev"

apiVersion: solutions.metalk8s.scality.com/v1alpha1
kind: Solution
metadata:
  annotations:
    solutions.metalk8s.scality.com/build-host: Latitude-E7450
    solutions.metalk8s.scality.com/build-timestamp: '2020-05-19T14:51:43Z'
    solutions.metalk8s.scality.com/development-release: '0'
    solutions.metalk8s.scality.com/display-name: Example Solution
    solutions.metalk8s.scality.com/git-revision: 2.4.3-104-g893792ca
  labels: {}
  name: example-solution
spec:
  images:
  - base-server:0.1.0-dev
  - example-solution-operator:0.1.0-dev
  - example-solution-ui:0.1.0-dev
  operator:
    image:
      name: example-solution-operator
      tag: 0.1.0-dev
  ui:
    image:
      name: example-solution-ui
      tag: 0.1.0-dev
  version: 0.1.0-dev

Refs: #2422

alexandre-allard committed:
example: generate manifest when building Solution

We now generate the new `manifest.yaml` instead of
the `product.txt` + `config.yaml` files with all the
content of both files in a new format as follows:

This change is done to simplify Solution archives,
with only a single file to describe the whole
Solution.

Also, the `config.yaml` was poorly named as it is not
a configuration file, but more a manifest.
`SolutionConfig` has been renamed to `Solution` to avoid
ambiguity with `SolutionsConfiguration` kind.

Refs: #2422

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2422-solutions-manifest` into
`development/2.4`
* [w/2.5/improvement/2422-solutions-manifest](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/2422-solutions-manifest) will be merged into `development/2.5` 
* [w/2.6/improvement/2422-solutions-manifest](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/2422-solutions-manifest) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

gdemonet commented on salt/_modules/metalk8s_solutions.py:
```diff
@@ -250,6 +251,58 @@ def read_solution_config(mountpoint, name, version):
+            )
+        )
+
+    manifest = yaml.safe_load(result['stdout'])
```
We should catch YAML errors here and re-raise as `CommandExecutionError`

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

TeddyAndrieux commented on salt/_modules/metalk8s_solutions.py:
```diff
@@ -202,52 +201,110 @@ def list_solution_images(mountpoint):
+    info = {}
+
+    try:
+        info['name'] = manifest.get('metadata', {})['name']
+        info['version'] = manifest.get('spec', {})['version']
```
Minor: Since you check for key error to me you do not want to use `get` here

alexandre-allard replied:
Yes, right, will keep it as is if it's the only requested change.

TeddyAndrieux reviewed (commented):
LGTM (just a minor comment but feel free to ignore it)

TeddyAndrieux reviewed (dismissed):
(ups wanted to approve)

gdemonet commented on salt/_modules/metalk8s_solutions.py:
```diff
@@ -202,52 +201,110 @@ def list_solution_images(mountpoint):
+
+
+def archive_info_from_iso(path):
+    """Extract archive information from a Solution ISO
```
We could call this method `manifest_from_iso` now that we have a good name for it :)

gdemonet commented on salt/_modules/metalk8s_solutions.py:
```diff
@@ -202,52 +201,110 @@ def list_solution_images(mountpoint):
+
+
+def _archive_info_from_manifest(manifest):
+    info = {}
```
As a general comment, this way of writing methods that build and return an object, although widespread in Salt modules, makes the code quite hard to read. You need to parse through each branch to understand what the contents of an object will be when it's returned. Sprinkle some early returns in there and it's quickly becoming a mess.

Instead, I'd suggest we only build the structure in the final `return` statement (even in an early `return`?). Something like this:

```python
def _archive_info_from_manifest(manifest):
    name = manifest.get('metadata', {}).get('name')
    version = manifest.get('spec', {}).get('version')

    if any(key is None for key in [name, version]):
        raise CommandExecutionError(
            'Missing mandatory key(s) in Solution "{}": must provide '
            '"metadata.name" and "spec.version"'
            .format(SOLUTION_MANIFEST)
        )

    display_name = manifest.get('annotations', {}).get(
        'solutions.metalk8s.scality.com/display-name', name
    )

    return {
        'name': name,
        'version': version,
        'display_name': display_name,
        'id': '-'.join([name, version]),
    }
```

alexandre-allard replied:
As you prefer, I did this because as you said, most of the time it's the approach used in Salt modules, so I tried to be consistent.

gdemonet replied:
Consistently bad is still bad :P I'd rather have us change this "standard" progressively by checking for this pattern in reviews than do a full refactor on thousands of lines in a few months...

gdemonet commented on salt/_modules/metalk8s_solutions.py:
```diff
@@ -202,52 +201,110 @@ def list_solution_images(mountpoint):
+    info = {}
+
+    try:
+        info['name'] = manifest.get('metadata', {})['name']
```
Although this comes from `metadata.name` in the manifest, this isn't an actual CRD (yet), so we need to perform validation manually and ensure this is lowercased and URL-encoded (because we'll use it, through the computed `info['id']`, as a path in our custom registry).

gdemonet commented on examples/metalk8s-solution-example/manifest.py:
```diff
@@ -0,0 +1,114 @@
+        'annotations': {},
+        'labels': {},
+    },
+    'spec': {
```
The list of extra images could be added as an option to the script, so we can check that overriding the default behaviour works as expected (we could build an ISO with a wrong manifest and check how our tooling behaves). But not needed for now, so feel free to leave this as debt (with a ticket to track and test) :)

alexandre-allard replied:
Added the list, so we just have to add the test now.
I'll create a ticket.

gdemonet replied:
ping

alexandre-allard replied:
#2618

gdemonet reviewed (changes requested):
Outside of the validation of the Solution `name` field, only a few minor comments.

Two more troubling questions:
- in 2.4? not sure that's a good strategy, it's a breaking change (would put it in 2.5.1 at the earliest)
- what about docs? I'd like to see at least an updated Solutions guide with an explanation and example of this new format, maybe even an extra note in the `CHANGELOG.md` warning about this change

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

alexandre-allard commented:
> Outside of the validation of the Solution `name` field, only a few minor comments.
> 
> Two more troubling questions:
> 
> * in 2.4? not sure that's a good strategy, it's a breaking change (would put it in 2.5.1 at the earliest)
> * what about docs? I'd like to see at least an updated Solutions guide with an explanation and example of this new format, maybe even an extra note in the `CHANGELOG.md` warning about this change

@gdemonet I wanted to put it in 2.4 to avoid Solution developers to maintain 2 format to support 2.4 and 2.5+, but maybe that's not what we want and they should just not use 2.4.

For the doc part... I would say that I use the "Code is the new Doc" or "Doc As Code" approach.
Joke aside, I agree, I will add some and also a note in the CHANGELOG.

bert-e commented:
# History mismatch



Merge commit #fbf9457bfe3d1cba1cfc8895316622f5ed005507 on the integration branch
`w/2.5/improvement/2422-solutions-manifest` is merging a branch which is neither the current
branch `improvement/2422-solutions-manifest` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/2422-solutions-manifest` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard committed:
docs: update Solutions architecture doc for new manifest.yaml

Replace mentions to `product.txt` and `config.yaml` with
the new `manifest.yaml` and adapt the doc.

Refs: #2422

alexandre-allard committed:
docs: update Solution developer guidelines

Various updates such as:
* Solution manifest
* Operator roles
* Solution interaction diagram
* Stuff from the Solution framework refactor
* Few fixes

Refs: #2422

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/2422-solutions-manifest` has diverged from
`development/2.5` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.5` into `improvement/2422-solutions-manifest`
* Rebase `improvement/2422-solutions-manifest` onto `origin/development/2.5`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# History mismatch



Merge commit #fbf9457bfe3d1cba1cfc8895316622f5ed005507 on the integration branch
`w/2.6/improvement/2422-solutions-manifest` is merging a branch which is neither the current
branch `improvement/2422-solutions-manifest` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/2422-solutions-manifest` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2422-solutions-manifest` into
`development/2.5`
* [w/2.6/improvement/2422-solutions-manifest](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/2422-solutions-manifest) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

miniscruff commented:
I was wondering if I should/need to adjust my PR here https://github.com/scality/Zenko/pull/1018 or if the versions expected / required make it not an issue. ( It is not a big change regardless but not sure of order or release timings ) @rahulreddy FYI.

bert-e commented:
# History mismatch



Merge commit #1c0a38c7f40a89cc6754fbf40d146809cd2f8b0e on the integration branch
`w/2.6/improvement/2422-solutions-manifest` is merging a branch which is neither the current
branch `improvement/2422-solutions-manifest` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/2422-solutions-manifest` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2422-solutions-manifest` into
`development/2.5`
* [w/2.6/improvement/2422-solutions-manifest](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/2422-solutions-manifest) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

alexandre-allard commented:
> I was wondering if I should/need to adjust my PR here [scality/Zenko#1018](https://github.com/scality/Zenko/pull/1018) or if the versions expected / required make it not an issue. ( It is not a big change regardless but not sure of order or release timings ) @rahulreddy FYI.

@miniscruff Yes, it would be better if you could adjust your PR accordingly, otherwise you'll have to manage different versions/formats.
There is also another change coming for the Operator roles (#2576).

alexandre-allard committed:
Add a CHANGELOG entry for Solutions manifest

Refs: #2422

alexandre-allard committed:
Fix documentation URL in the CHANGELOG.md

gdemonet commented on CHANGELOG.md:
```diff
@@ -2,6 +2,15 @@
+  ([#2422](https://github.com/scality/metalk8s/issues/2422)).
+  See [Solutions documentation](https://metal-k8s.readthedocs.io/en/development-2.5/developer/solutions/archive.html#product-information)
+  for details about the new format.
+
```
Can you add a statement explaining the impact? Like `Solution archives working on previous versions of MetalK8s will no longer be compatible and will need to be regenerated` or something...

gdemonet commented on docs/developer/architecture/solutions.rst:
```diff
@@ -116,35 +116,32 @@ Here is how it will look::
+        image:
+          name: solution-name-operator
+          tag: 1.0.0
+      ui:
+        image:
+          name: solution-name-ui
+          tag: 1.0.0
```
So we're still deploying UIs with MetalK8s... Let's open a debt ticket to remove this behaviour. Also, we may want to add some docs explaining that if this section is not specified, we won't try to deploy the UI ourselves (recommended approach, let the Operator do it).

alexandre-allard replied:
Ok, opened the ticket #2581

gdemonet commented on docs/developer/solutions/archive.rst:
```diff
@@ -67,56 +67,56 @@ Here is the file tree expected by MetalK8s to exist in each Solution archive::
+       image:
+         name: solution-name-operator
+         tag: 1.0.0
+     ui:
+       image:
+         name: solution-name-ui
+         tag: 1.0.0
```
Maybe remind that this is optional and conditions the deployment of said UI by MetalK8s.

alexandre-allard replied:
Yep

gdemonet commented on docs/developer/solutions/archive.rst:
```diff
@@ -67,56 +67,56 @@ Here is the file tree expected by MetalK8s to exist in each Solution archive::
+     $(date -u +%Y-%m-%dT%H:%M:%SZ)
+   solutions.metalk8s.scality.com/git-revision: \
+     $(git describe --always --long --tags --dirty)
 
```
Maybe add a note linking to your simple script for generating the manifest? Just in case readers want to copy-paste it.

gdemonet commented on docs/developer/solutions/introduction.rst:
```diff
@@ -28,49 +27,48 @@ For more details, see the following documentation pages:
-into separate namespaces. Using the Operator-defined ``CustomResource(s)``, the
-user can then effectively deploy the application packaged in the Solution.
+Once a Solution is imported in MetalK8s, a user can deploy one or more versions
+of the Solution Operator, using either the Solution UI or CLI, into separate
```
`Solution UI or CLI`: what's that? I think you mean the `solutions.sh` script or the "Environments" page in MetalK8s UI?
Interacting with the CRs would then use a `Solution UI` or `CLI` (probably kubectl plugins).

alexandre-allard replied:
Yes, that's what I meant, I'll update this to make it more clear.

gdemonet commented on docs/developer/solutions/operator.rst:
```diff
@@ -69,28 +56,38 @@ to set up monitoring of your application services.
+as the endpoints for the container images.
+The given configuration looks like::
+
+   apiVersion: solutions.metalk8s.scality.com/v1alpha1
+   kind: OperatorConfig
+   repositories:
+     <solution-version-x>:
+       - endpoint: metalk8s-registry/<solution-name>-<solution-version-x>
+         images:
+           - <image-x>:<tag-x>
+           - <image-y>:<tag-y>
+     <solution-version-y>:
+       - endpoint: metalk8s-registry/<solution-name>-<solution-version-y>
+         images:
+           - <image-x>:<tag-x>
+           - <image-y>:<tag-y>
```
Cool, but not really explaining why we use this format or what it enables them to do with it.
To make this a bit more obvious, you could show this format as "what MetalK8s will inject when deploying your Operator", and show another example of what the Operator could use as defaults when running with online access, e.g.:

```yaml
apiVersion: solutions.metalk8s.scality.com/v1alpha1
kind: OperatorConfig
repositories:
  1.0.0:
    - endpoint: registry.scality.com/zenko
      images:
        - cloudserver:1.0.0
        - zenko-quorum:1.0.0
    - endpoint: quay.io/coreos
      images:
        - prometheus-operator:v0.34.0
```

gdemonet commented on docs/developer/solutions/operator.rst:
```diff
@@ -69,28 +56,38 @@ to set up monitoring of your application services.
-.. _`Operator Lifecycle Manager`:
-   https://github.com/operator-framework/operator-lifecycle-manager
+   ``Role`` and ``RoleBinding`` should be preferred over ``ClusterRole`` and
+   ``ClusterRoleBinding``, if there is no need for cluster-scoped permissions.
```
Would be nice to track this topic in a ticket: most importantly, we don't want to allow cluster-scoped permissions for Environment-scoped Operators. But maybe we'll need other scopes, and the question of multiple namespaces within an environment is also non-trivial, so let's tackle it later.

gdemonet replied:
ping

alexandre-allard replied:
#2619

gdemonet commented on examples/metalk8s-solution-example/manifest.py:
```diff
@@ -0,0 +1,133 @@
+    """Script entrypoint."""
+    args = parse_arguments()
+
+    manifest = SOLUTION_MANIFEST
```
FYI: if it's something you're mutating in your method, then it can't be a constant. Hence, can't be using the constants naming conventions, to avoid confusing a user.

If the intention is to generate a new object from a set of args, then just define this object in the method body. This way, there's no risk of side effects:

```python
def _img_dict(name, tag):
    return {'image': {'name': name, 'tag': tag}}

def build_manifest(args):
    return {
        'apiVersion': 'solutions.metalk8s.scality.com/v1alpha1',
        'kind': 'Solution',
        'metadata': {
            'name': args.name,
            'annotations': dict(args.annotations),
            'labels': dict(args.labels),
        },
        'spec': {
            'images': list(set(
                ':'.join(image) for image in (
                    args.extra_images + [args.ui_image, args.operator_image]
                )
            )),
            'ui': _img_dict(*args.ui_image),
            'operator': _img_dict(*args.operator_image),
            'version': args.version,
        },
    }

# Try to make your functions testable
# Here, we can pass in a stream to output the manifest string,
# so in a test we can just pass an in-memory buffer and inspect it
def main(stream=sys.stdout):
    args = parse_arguments()
    manifest = build_manifest(args)

    # Also, no need to manually write,
    # just pass the stream to `yaml.safe_dump`
    yaml.safe_dump(manifest, stream, default_flow_style=False)
```

gdemonet replied:
It's just a helper script, so no need to change anything. But it's not longer to write code "the right way", so I hope you'll think about it in the future ;)

alexandre-allard replied:
Yes, I know this, but since it's just a simple script to generate a YAML, I thought it would be sufficient.
Oh cool for the stream stuff, didn't know about that one.

gdemonet reviewed (changes requested):
We have some time before releasing 2.5.1 I think, so let's improve a bit more :)

bert-e commented:
# History mismatch



Merge commit #1dc0c2c30f5f049763c625b6c9341bc64dea526a on the integration branch
`w/2.6/improvement/2422-solutions-manifest` is merging a branch which is neither the current
branch `improvement/2422-solutions-manifest` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/2422-solutions-manifest` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2422-solutions-manifest` into
`development/2.5`
* [w/2.6/improvement/2422-solutions-manifest](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/2422-solutions-manifest) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

gdemonet commented on salt/metalk8s/orchestrate/solutions/prepare-environment.sls:
```diff
@@ -155,8 +154,8 @@ Cannot deploy Solution {{ name }}-{{ version }} for environment {{ env_name }}:
-              {{- deploy_operator(namespace, solution) }}
-              {{- deploy_admin_ui(namespace, solution) }}
+              {{- deploy_operator(namespace, name, solution) }}
+              {{- deploy_admin_ui(namespace, name, solution) }}
```
Maybe calling this macro should be made optional based on the presence of a `spec.ui.image` key in the manifest? Or we can just ignore until #2581 is done

gdemonet reviewed (approved):
LGTM!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

alexandre-allard commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
